utils/forms/forms.js

```diff
@@ -0,0 +1,368 @@
+  } else {
+    // Clear existing error messages if the container already exists
+    errorContainer.innerHTML = '';
+  }
+
+  // Append each error message as a separate span element
+  errorMessages.forEach((msg) => {
+    const span = document.createElement('span');
+    span.className = 'forms-error';
+    span.textContent = msg;
+    errorContainer.append(span);
+  });
+}
+
+function clearError(input) {
+  const existingError = input.nextElementSibling;
+  if (existingError && existingError.classList.contains('error-messages')) {
+    existingError.remove();
+  }
+}
+
+function validateInput(input) {
+  input.setCustomValidity('');
+  const errorMessages = [];
+  let isValid = true;
+
+  // Run built-in validation and gather messages
+  if (input.validationMessage) {
+    errorMessages.push(input.validationMessage);
+  }
+
+  // Retrieve and parse rules
+  const validationRules = JSON.parse(input.dataset.validation || '[]');
+  validationRules.forEach((rule) => {
+    if (rule === 'no-nums' && /\d/.test(input.value)) {
+      errorMessages.push('Numbers are not allowed.');
+    }
+
+    if (rule === 'phone:US' && /\d/.test(input.value)) {
+      const digitsOnly = input.value.replace(/\D/g, '');
+
+      if (digitsOnly.length < 10) {
+        errorMessages.push(`Missing ${10 - digitsOnly.length} digit${digitsOnly.length === 9 ? '' : 's'}`);
+      }
+
+      if (digitsOnly.length > 10) {
+        errorMessages.push('There should only be 10 digits in this entry.');
+      }
+    }
+  });
+
+  // Set custom validity if there are any errors
+  if (errorMessages.length > 0) {
+    isValid = false;
+    input.setCustomValidity(errorMessages.join('\n')); // This will trigger :invalid
+    showError(input, errorMessages);
+  } else {
+    clearError(input);
+  }
+
+  return isValid;
+}
+
+function validateForm(form) {
+  let isValid = true;
+
+  // Loop through all form inputs
+  const formInputs = Array.from(form.elements);
+  formInputs.forEach((input) => {
+    // Skip non-input elements like buttons
+    if (input.type !== 'submit' && input.type !== 'button') {
+      const inputValid = validateInput(input);
+      if (!inputValid) {
+        isValid = false; // Mark the form as invalid if any field fails
+      }
+    }
+  });
+
+  return isValid;
+}
+
+function buildInput(field) {
+  const input = document.createElement('input');
+  input.type = field.type || 'text';
+  input.name = field.name || '';
+  input.placeholder = field.placeholder || '';
+
+  if (field.value !== undefined && field.value !== null) {
+    input.value = field.value;
+  }
+
+  if (field.min !== undefined && field.min !== null) {
+    input.min = field.min;
+  }
+
+  // Apply additional properties if they exist
+  if (field.required) input.required = true;
+  if (field.minLength) input.minLength = field.minLength;
+  if (field.maxLength) input.maxLength = field.maxLength;
+  if (field.max) input.max = field.max;
```

Comment:
`min` and `minLength`?

Reply:
Min is for number inputs and minLength is for password inputs